<!DOCTYPE html>
<title>Database Storage - Nested transactions</title>
<pre id="result">FAIL - Script did not execute</pre>

<script>
//Database openDatabase(in DOMString name, in DOMString version, in DOMString displayName, in unsigned long estimatedSize);
var result = document.getElementById("result");
result.innerHTML = "";

function record(pass, msg) {
	result.innerHTML += (pass ? "PASS" : "FAIL") + (msg ? " - " + msg : "") + "\n";
	try{top.opener.rr(pass);}catch(e){}
}

function error() {
	e = arguments.length == 1 ? arguments[0] : arguments[1];
	record(false, "Error " + e.code + ": " + e.message);
}

try {
	opera.deleteDatabase("test107");
	var db = window.openDatabase("test107", "", "Test DB", "100000");

	function initialData(transaction) {
		//void executeSql(sqlStatement, [arguments], callback, errorCallback);
		transaction.executeSql("DROP TABLE IF EXISTS t;", [], null, error);
		transaction.executeSql("CREATE TABLE t(id INTEGER, label TEXT);", [], null, error);
		transaction.executeSql("INSERT INTO t VALUES (?,?)", ['1', 'one'], null, error);
	}

	function innerTransactionCallback(transaction) {
		transaction.executeSql("INSERT INTO t VALUES (?, ?)", [2, 'two'], null, error);
	}

	function outerTransactionCallback(transaction) {
		transaction.executeSql("INSERT INTO t VALUES (?, ?)", [99, 'nines'], null, error);
		db.transaction(innerTransactionCallback, error);
		transaction.executeSql("INSERT INTO t VALUES (?, ?)", [999, 'more nines'], null, error);
		throw "die!";
	}

	function checkData () {
		db.transaction(function(transaction) {
			transaction.executeSql("SELECT * FROM t ORDER BY id", [], function(transaction, resultSet) {
				try {
					if (resultSet.rows.length === 2 && resultSet.rows.item(0).id === 1 && resultSet.rows.item(1).id === 2) {
						// even though the outer callback fails the inner one completes successfully
						record(true);
					} else {
						record(false, "Wrong results");
					}
				} catch (e) {
					record(false, "Exception while checking results: " + e);
				}
			});
		});
	}

	db.transaction(initialData, error, db.transaction(outerTransactionCallback, checkData, function(){record(false,"sanity check: should not call success callback")}));
} catch (e) {
	record(false, "Couldn't create database or complete transaction: " + e);
}
</script>
